components/org.wso2.carbon.identity.conditional.auth.functions.choreo/src/main/java/org/wso2/carbon/identity/conditional/auth/functions/choreo/CallChoreoFunctionImpl.java

```diff
@@ -0,0 +1,173 @@
+                                    JSONParser parser = new JSONParser();
+                                    json = (JSONObject) parser.parse(jsonString);
+                                    outcome = Constants.OUTCOME_SUCCESS;
+                                } catch (ParseException e) {
+                                    LOG.error("Error while building response from Choreo call for " +
+                                            "session data key: " + authenticationContext.getContextIdentifier(), e);
+                                    outcome = Constants.OUTCOME_FAIL;
+                                } catch (IOException e) {
+                                    LOG.error("Error while reading response from Choreo call for " +
+                                            "session data key: " + authenticationContext.getContextIdentifier(), e);
+                                    outcome = Constants.OUTCOME_FAIL;
+                                }
+                            } else {
+                                outcome = Constants.OUTCOME_FAIL;
+                            }
+                            asyncReturn.accept(authenticationContext, json != null ? json : Collections.emptyMap(),
+                                    outcome);
+                        } catch (FrameworkException e) {
+                            LOG.error("Error while proceeding after successful response from Choreo call for "
+                                    + "session data key: " + authenticationContext.getContextIdentifier(), e);
+                        }
+                    }
+
+                    @Override
+                    public void failed(final Exception ex) {
+
+                        LOG.error("Failed to invoke Choreo for session data key: " +
+                                authenticationContext.getContextIdentifier(), ex);
+                        try {
+                            if (LOG.isDebugEnabled()) {
+                                LOG.debug(" Calls to Choreo failed for session " +
+                                        "data key: " + authenticationContext.getContextIdentifier());
+                            }
+                            String outcome = OUTCOME_FAIL;
+                            if ((ex instanceof SocketTimeoutException)
+                                    || (ex instanceof ConnectTimeoutException)) {
+                                outcome = OUTCOME_TIMEOUT;
+                            }
+                            asyncReturn.accept(authenticationContext, Collections.emptyMap(), outcome);
+                        } catch (FrameworkException e) {
+                            LOG.error("Error while proceeding after failed response from Choreo " +
+                                    "call for session data key: " + authenticationContext
+                                    .getContextIdentifier(), e);
```

Comment:
shouldn't we accept the async return with a failed outcome?

Reply:
This catch block handles exceptions on asyncReturn.accept() method.